### ISSUE TITLE
Serving a single file for multiple routes

### ISSUE THREAD

**Issue by Krasjet** [haskell-servant/servant#1281]
What should be the best way to serve a single static file (HTML file, in particular) for multiple routes?

I know there is `serveDirectoryWebApp` for serving an entire directory, but it can't handle serving multiple routes with the same file.

I'm currently loading the file using `readFile` then return it as a Blaze Html object. The code roughly looks like this
```haskell
import qualified Data.Text.IO as TIO
import Control.Monad.IO.Class  (liftIO)
import Servant
import Servant.HTML.Blaze
import Text.Blaze.Html

type EditDoc = "edit" :> Capture "docId" DocId :> Get '[HTML] Html

-- | Send the editor page
editDoc :: DocId ->  Handler Html
editDoc docId = do
  -- [snip] check if docId is valid
  html <- liftIO $ TIO.readFile $ "editor.html"
  return $ preEscapedToMarkup html
```
but I don't think this is efficient. First, there is no header attached to it. The response header of this approach looks like this
```
HTTP/1.1 200 OK
Content-Type: text/html;charset=utf-8
Date: Mon, 16 Mar 2020 18:51:47 GMT
Server: Warp/3.3.5
Transfer-Encoding: chunked
```
but with `serveDirectoryWebApp`, there is some extra cache information attached to it
```
HTTP/1.1 200 OK
Accept-Ranges: bytes
Cache-Control: public, max-age=31536000
Content-Length: 1144
Content-Type: text/html
Date: Mon, 16 Mar 2020 18:52:01 GMT
ETag: UNStdlqW7HoO2fGDt79I2w==
Expires: Thu, 31 Dec 2037 23:55:55 GMT
Last-Modified: Mon, 16 Mar 2020 18:19:26 GMT
Server: Warp/3.3.5
```
Also, wai has a native [`responseFile`](https://hackage.haskell.org/package/wai-3.2.2.1/docs/Network-Wai.html#v:responseFile) function for sending files so you can write something like this (stole from [here](https://hackage.haskell.org/package/wai))
```
plainIndex :: Response
plainIndex = responseFile
    status200
    [("Content-Type", "text/plain")]
    "index.html"
    Nothing
```
and warp seems to implement it with the [sendfile()](https://github.com/yesodweb/wai/blob/b29e2705d0b4035a17227a8669c6430eb7c9d067/warp/Network/Wai/Handler/Warp/SendFile.hs#L35-L51) syscall, which can be more efficient than reading and writing.

Is there a better way to handle this?

**Comment by alpmestan** [comment]
I'd probably recommend using the same handler for all your routes, and making them all `Raw` endpoints so as to be able to use `responseFile` directly. If the file is small, you could alternatively read it before starting the server and then just return its contents directly whenever it is requested.

**Comment by Krasjet** [comment]
I spent some time to figure out how to use the `Raw` route with custom monads, but the minimal working example would be like this:

```haskell
import Network.HTTP.Types
import Network.Wai
import Servant

type EditDoc = "edit" :> Capture "docId" DocId :> Raw

-- | Send the editor page
editDoc :: DocId -> Server Raw
editDoc docId = Tagged $ \req res ->
  -- [snip] check if docId is valid
  res $ responseFile status200 [("Content-Type", "text/html")] "editor.html" Nothing
```

**Comment by tysonzero** [comment]
I'm running into a similar situation, and I have to say it'd be convenient if there was just a simple helper function for it in `Servant.Server.StaticFiles`.

```
serveFileWebApp :: FilePath -> ServerT Raw m
```

**Comment by alpmestan** [comment]
That seems reasonable, although this function would necessarily re-read the file on every request, whereas some use cases would perhaps benefit from reading the file beforehand and serving its contents from memory. Perhaps some file-embed style alternative would be good there. Discussions/patches are certainly welcome :-)

**Comment by ad-si** [comment]
Shall we open a new Issue for a `serveFile` function as described by @tysonzero?
I also think this would be a good idea!
